pkg/types/types.go

```diff
@@ -234,52 +244,91 @@ func (s byStepTime) Less(i, j int) bool {
 	return s[i].StepTime < s[j].StepTime
 }
 
-func mergeMetrics(metrics []Metric) Metric {
+func getPointInconsistentValueCount(selectedValue float64, values []float64) int {
+	valuesCount := len(values)
+	if valuesCount == 0 {
+		// No value for the point
+		return 0
+	}
+
+	valuesToCount := make(map[float64]int)
+	for _, val := range values {
+		valuesToCount[val]++
+	}
+
+	return valuesCount - valuesToCount[selectedValue]
+}
+
+func mergeMetrics(metrics []Metric, consistencyCheck bool) (metric Metric, inconsistencies int) {
 	if len(metrics) == 0 {
-		return Metric{}
+		return Metric{}, 0
 	}
 
 	if len(metrics) == 1 {
-		return metrics[0]
+		m := metrics[0]
+		return m, 0
 	}
 
 	sort.Sort(byStepTime(metrics))
 	healed := 0
 
 	// metrics[0] has the highest resolution of metrics
-	metric := metrics[0]
+	metric = metrics[0]
 	for i := range metric.Values {
-		if !metric.IsAbsent[i] {
-			continue
+		var valuesForPoint []float64
+		pointExists := !metric.IsAbsent[i]
+		if pointExists {
+			valuesForPoint = append(valuesForPoint, metric.Values[i])
 		}
-
-		// found a missing value, look for a replacement
 		for j := 1; j < len(metrics); j++ {
+			if pointExists && !consistencyCheck {
+				break
+			}
 			m := metrics[j]
 
 			if m.StepTime != metric.StepTime || len(m.Values) != len(metric.Values) {
 				break
 			}
 
-			// found one
-			if !m.IsAbsent[i] {
+			if m.IsAbsent[i] {
+				continue
+			}
+
+			valuesForPoint = append(valuesForPoint, m.Values[i])
+
+			if !pointExists {
 				metric.IsAbsent[i] = m.IsAbsent[i]
 				metric.Values[i] = m.Values[i]
 				healed++
-				break
+				pointExists = true
+			}
+		}
+		if consistencyCheck {
+			dataPointInconsistentValues := getPointInconsistentValueCount(metric.Values[i], valuesForPoint)
+			if dataPointInconsistentValues > 0 {
+				inconsistencies++
 			}
 		}
 	}
 
+	if consistencyCheck && inconsistencies > 0 {
+		corruptionLogger.Warn("metric consistency issue",
```

Comment:
This can potentially write a log of logs. Two options to solve this:
1. Put rate-limiting on the logger.
2. Log per-request and include a truncated list of inconsistent metrics in the log.

Reply:
chose 2.